llm/server.go

```diff
@@ -0,0 +1,643 @@
+		return nil, fmt.Errorf("no servers found for %v", info)
+	}
+
+	dir := available[servers[0]]
+
+	// TODO: let user override with OLLAMA_LLM_LIBRARY
+	slog.Info("using server", "server", servers[0])
+
+	params := []string{
+		"--model", model,
+		"--ctx-size", fmt.Sprintf("%d", opts.NumCtx),
+		"--batch-size", fmt.Sprintf("%d", opts.NumBatch),
+		"--embedding",
+		"--log-disable",
+	}
+
+	if opts.NumGPU > 0 {
+		params = append(params, "--n-gpu-layers", fmt.Sprintf("%d", opts.NumGPU))
+	}
+
+	if debug := os.Getenv("OLLAMA_DEBUG"); debug != "" {
+		fmt.Println("adding verbose")
+		params = append(params, "--verbose")
+	}
+
+	if opts.MainGPU > 0 {
+		params = append(params, "--main-gpu", fmt.Sprintf("%d", opts.MainGPU))
+	}
+
+	if opts.RopeFrequencyBase > 0 {
+		params = append(params, "--rope-freq-base", fmt.Sprintf("%f", opts.RopeFrequencyBase))
+	}
+
+	if opts.RopeFrequencyScale > 0 {
+		params = append(params, "--rope-freq-scale", fmt.Sprintf("%f", opts.RopeFrequencyScale))
+	}
+
+	if len(adapters) > 0 {
+		// TODO: applying multiple adapters is not supported by the llama.cpp server yet
+		params = append(params, "--lora", adapters[0])
+	}
+
+	if len(projectors) > 0 {
+		// TODO: applying multiple projectors is not supported by the llama.cpp server yet
+		params = append(params, "--mmproj", projectors[0])
+	}
+
+	if opts.NumThread > 0 {
+		params = append(params, "--threads", fmt.Sprintf("%d", opts.NumThread))
+	}
+
+	if !opts.F16KV {
+		params = append(params, "--memory-f32")
+	}
+
+	if opts.UseMLock {
+		params = append(params, "--mlock")
+	}
+
+	if !opts.UseMMap {
+		params = append(params, "--no-mmap")
+	}
+
+	if opts.UseNUMA {
+		params = append(params, "--numa")
+	}
+
+	port := rand.Intn(65535-49152) + 49152 // get a random port in the ephemeral range
+	params = append(params, "--port", strconv.Itoa(port))
+
+	slog.Info("starting llama server", "params", params)
+
+	// append the server directory to LD_LIBRARY_PATH
+	var libraryPaths []string
+	if libraryPath, ok := os.LookupEnv("LD_LIBRARY_PATH"); ok {
+		libraryPaths = append(libraryPaths, libraryPath)
+	}
+	libraryPaths = append(libraryPaths, dir)
+
+	server := filepath.Join(dir, "ollama_llama_server")
+	if runtime.GOOS == "windows" {
+		server = server + ".exe"
+	}
+
+	s := &LlamaServer{
+		port: port,
+		cmd:  exec.Command(server, params...),
+	}
+
+	s.cmd.Env = append(os.Environ(), fmt.Sprintf("LD_LIBRARY_PATH=%s", strings.Join(libraryPaths, ":")))
+	s.cmd.Stdout = os.Stdout
+	s.cmd.Stderr = os.Stderr
+
+	slog.Info("starting llama", "cmd", s.cmd.String())
+
+	if err := s.cmd.Start(); err != nil {
+		return nil, fmt.Errorf("error starting the external llama server: %v", err)
+	}
+
+	go func() {
+		s.done <- s.cmd.Wait() // This will capture the exit error
+	}()
+
+	if err := s.waitUntilRunning(); err != nil {
+		log.Printf("error starting llama server: %v", err)
+		s.Close()
+
+		return nil, err
+	}
+
+	return s, nil
+}
+
+func (s *LlamaServer) Wait() <-chan error {
```

Comment:
This isn't used yet, although it should be to detect startup errors

Reply:
I see this is in your `TODO` but just wanted to leave a comment that this is super important. I'm trying to improve error reporting on our current implementation's start-up right now, it is becoming a bit of a problem how many errors are returned with the same message.